node/beacon/src/authority.rs

```diff
@@ -81,51 +90,106 @@ impl Authority {
             ..=(epoch + 1) * self.authority_config.epoch_length // Without ..= it needs + 1.
     }
 
+    fn read_from_storage(&mut self) {
+        let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
+        let proposals = guard.get_proposal();
+        if !proposals.is_empty() {
+            self.proposals = proposals;
+            let participation = guard.get_participation();
+            self.participation = participation;
+            let processed_blocks = guard.get_processed_blocks();
+            self.processed_blocks = processed_blocks;
+            let thresholds = guard.get_threshold();
+            self.thresholds = thresholds;
+            let accepted_authorities = guard.get_accepted_authorities();
+            self.accepted_authorities = accepted_authorities;
+        }
+    }
+
+    fn prune_slot<V: Clone>(&self, data: &HashMap<Slot, V>) -> HashMap<Slot, V> {
+        data.iter()
+            .filter(|(&k, &_)| self.slot_to_epoch(k) >= self.current_epoch)
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .collect()
+    }
+
+    fn prune_epoch<V: Clone>(&self, data: &HashMap<Epoch, V>) -> HashMap<Epoch, V> {
+        data.iter()
+            .filter(|(&k, &_)| k >= self.current_epoch)
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .collect()
+    }
+
+    fn write_to_storage(&self) {
+        // prune old epochs
+        let proposals = self.prune_slot(&self.proposals);
```

Comment:
should we also update self.proposals/participation/etc here?
e.g. as we write, we actually prune the old data as well in memory?

Reply:
Fixed

Reply:
Looks like we cannot prune this right now because of the hard coded block index in protocol https://github.com/nearprotocol/nearcore/blob/29957183f8d60310710f9a455b08c882f73ad4af/node/network/src/protocol.rs#L162